syntax = "proto3";

import "flavors.proto";

package instances;

option cc_generic_services = true;


service InstanceGrpcService {
    rpc list_instances (ListInstancesRequest) returns (ListInstancesResponse) {}
    rpc create_instance (CreateInstanceRequest) returns (CreateInstanceResponse) {}
    rpc delete_instance (DeleteInstanceRequest) returns (DeleteInstanceResponse) {}
    rpc start_instance (StartInstanceRequest) returns (InstanceResponse) {}
    rpc stop_instsance (StopInstanceRequest) returns (InstanceResponse) {}
    rpc AttachDisk(AttachDiskRequest) returns (InstanceResponse);
    rpc DetachDisk(DetachDiskRequest) returns (InstanceResponse);
    rpc attach_disk (AttachDiskRequest) returns (InstanceResponse) {}
    rpc dettach_disk (DetachDiskRequest) returns (InstanceResponse) {}
    rpc take_snapshot (TakeSnapshotRequest) returns (TakeSnapshotResponse) {}
    rpc export_development_image (ExportDevelopmentImageRequest) returns (ExportDevelopmentImageResponse) {}
}


message Instance {
    string name = 1;
    string image = 2;
    string vm_state = 3;
    string ip_address = 4;
    optional flavors.Flavor flavor = 5;
}

message ListInstancesRequest {
}


message ListInstancesResponse {
    repeated Instance instances = 1;
}

message CreateInstanceRequest {
    string name = 1;
    string image = 2;
    string arch = 3;
    string instance_path = 4;
    string root_disk_path = 5;
    flavors.Flavor flavor = 6;
}

message CreateInstanceResponse {
    uint32 ret = 1;
    string msg = 2;
    optional Instance instance = 3;
}

message DeleteInstanceRequest {
    string name = 1;
}

message DeleteInstanceResponse {
    uint32 ret = 1;
    string msg = 2;
}

message StartInstanceRequest {
    string name = 1;
}

message StopInstanceRequest {
    string name = 1;
}

message TakeSnapshotRequest {
    string name = 1;
    string snapshot = 2;
    string dest_path = 3;
}

message TakeSnapshotResponse {
    string msg = 1;
}

message ExportDevelopmentImageRequest {
    string name = 1;
    string image = 2;
    string dest_path = 3;
    string pwd = 4;
}

message ExportDevelopmentImageResponse {
    string msg = 1;
}

message AttachDiskRequest {
    string name = 1;
    string disk_path = 2;
    string drive_type = 3;
}

message DetachDiskRequest {
    string name = 1;
    string disk_path = 2;
}

message InstanceResponse {
    string message = 1;
}
